gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt

```diff
@@ -0,0 +1,129 @@
+package com.google.devtools.ksp.gradle
+
+import org.gradle.api.InvalidUserCodeException
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.jetbrains.kotlin.gradle.dsl.*
+import org.jetbrains.kotlin.gradle.plugin.*
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinJvmAndroidCompilation
+
+/**
+ * Creates and retrieves ksp-related configurations.
+ */
+class KspConfigurations(private val project: Project) {
+    companion object {
+        private const val PREFIX = "ksp"
+    }
+
+    // Store all ksp configurations for quick retrieval.
+    private val kotlinConfigurations = mutableMapOf<KotlinSourceSet, Configuration>()
+    private val androidConfigurations = mutableMapOf<String, Configuration>()
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun <T : Any> saveConfiguration(
+        ownerSet: T,
+        ownerName: String,
+        name: String,
+        cache: MutableMap<T, Configuration>
+    ): Configuration {
+        val configName = PREFIX + name.replaceFirstChar { it.uppercase() }
+        // maybeCreate to be future-proof, but we should never have a duplicate with current logic
+        val config = project.configurations.maybeCreate(configName).apply {
+            description = "KSP dependencies for the '$ownerName' source set."
+            isCanBeResolved = false // we'll resolve the processor classpath config
+            isCanBeConsumed = false
+            isVisible = false
+        }
+        cache[ownerSet] = config
+        return config
+    }
+
+    private fun saveKotlinConfiguration(owner: KotlinSourceSet, name: String) =
+        saveConfiguration(owner, owner.name, name, kotlinConfigurations)
+
+    private fun saveAndroidConfiguration(key: String, name: String) =
+        saveConfiguration(key, "$key (Android)", name, androidConfigurations)
+
+    init {
+        project.plugins.withType(KotlinBasePluginWrapper::class.java).configureEach {
+            // 1.6.0: decorateKotlinProject(project.kotlinExtension)?
+            decorateKotlinProject(project.extensions.getByName("kotlin") as KotlinProjectExtension, project)
+        }
+    }
+
+    private fun decorateKotlinProject(kotlin: KotlinProjectExtension, project: Project) {
+        when (kotlin) {
+            is KotlinSingleTargetExtension -> decorateKotlinTarget(kotlin.target)
+            is KotlinMultiplatformExtension -> {
+                kotlin.targets.configureEach(::decorateKotlinTarget)
+
+                // Adding multiplatform configuration removed support for the root ksp configuration.
+                // Try to make this breaking change less breaking by adding a clear error.
+                project.configurations.create("ksp").dependencies.all {
+                    throw InvalidUserCodeException(
```

Comment:
Pardon me if I missed any discussion in the review thread, but I'm wondering why taking this away. Isn't an apply-to-all configuration (except `kspTest`) useful?

Reply:
It was discussed here https://github.com/google/ksp/pull/609#discussion_r714226247 I don't have a strong opinion FWIW, up to you guys. To summarize I think Ivan point was that in KMP if you do

```kotlin
kotlin {
    jvm()
    sourceSets.jvmMain {
        dependencies { this: KotlinDependenciesHandler ->
            implementation("foo")
            project.dependencies.add("ksp", "bar")
        }
    }
}
```

then `foo` is jvm only, while `bar` will be applied to all targets, and this is confusing. I think we could add `KotlinDependenciesHandler.ksp()` DSL extensions so that people just write `ksp(bar)`, but I'm not sure about the implementation (need to cast KDH to some internal Kotlin plugin type) and also didn't want to make this PR bigger.

Reply:
To echo Mattia's point, I think that having a global `ksp` configuration is quite error prone. If there is a need for a configuration that adds to all platforms, it should probably be named `kspAllPlatforms` or something like that.

Reply:
Got it, thanks. I'm worrying about the breakage 😢 We are planning for 1.0.1 release and this PR is definitely the most important one to include. I'm wondering whether we should keep the universal `ksp` for now, and announce that we are going to remove it in a later release? So that people are less surprised and have a chance to migrate. As a mitigation, I can polish the doc, add an example of KMP and tell people to use `ksp<SourceSet>` instead of `ksp`.

What do you think?

Reply:
Agree, it does seem like the current approach may be a bit too aggressive. This migration can be more smooth by doing it in 3 stages:
- keep `ksp` and have other configurations extend from it but add warning that folks should migrate to source-set specific configurations
- keep `ksp` configuration, but make it a no-op, and tell users to migrate
- remove `ksp` configuration fully